effectful/handlers/pyro.py

```diff
@@ -252,190 +264,280 @@ def __repr__(self):
         return f"Naming({self.name_to_dim})"
 
 
-class PositionalDistribution(pyro.distributions.torch_distribution.TorchDistribution):
-    """A distribution wrapper that lazily converts indexed dimensions to
-    positional.
+@defop
+def named_distribution(
+    d: Annotated[TorchDistribution, Scoped[A | B]],
+    *names: Annotated[Operation[[], int], Scoped[B]],
+) -> Annotated[TorchDistribution, Scoped[A | B]]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            if dist_constr is dist.Independent:
```

Comment:
I don't think this special case handling of `Independent` is necessary. You just need to distribute `named_distribution` across the arguments:
```py
d = defterm(d) if not isinstance(d, Term) else d
if not (d.op is call and issubclass(d.args[0], TorchDistribution)):
  raise NotImplementedError

dist_constr, args = args[0], args[1:]

def _to_named(a):
  if isinstance(a, torch.Tensor):
    return Indexable(typing.cast(torch.Tensor, a))[tuple(n() for n in names)]
  elif isinstance(a, TorchDistribution):
    return named_distribution(a, *names)
  else:
    return a
 
return dist_constr(*tree.map_structure(_to_named, args))
```